### ISSUE TITLE
[add] libheif

### ISSUE THREAD

**Issue by zedd-0** [m-ab-s/media-autobuild_suite#2516]
hi, is it possible to add libheif to this script?
https://github.com/strukturag/libheif

**Comment by hydra3333** [comment]
I don't recall if ffmpeg supports those formats already in some manner ... 
Ah. It seems ffmpeg itself does not support heif
https://stackoverflow.com/questions/65435536/convert-a-heif-file-to-png-jpg-using-ffmpeg

After quick google, I have not yet spotted anything in ffmpeg to say it supports libheif and a reasonable guess would be that ffmpeg needs to be updated to support it.

In fact https://github.com/strukturag/libheif#software-using-libheif does not list ffmpeg as software which uses libheif.

On that basis, since MABS is a _build system_ for ffmpeg and dependencies, I guess that no change to MABS would achieve what you are seeking ?

Lack of heif support seems to be a long overdue shortcoming in ffmpeg, based on TRAC comments.

**Comment by L4cache** [comment]
> I don't recall if ffmpeg supports those formats already in some manner ... Ah. It seems ffmpeg itself does not support heif https://stackoverflow.com/questions/65435536/convert-a-heif-file-to-png-jpg-using-ffmpeg
> 
> After quick google, I have not yet spotted anything in ffmpeg to say it supports libheif and a reasonable guess would be that ffmpeg needs to be updated to support it.
> 
> In fact https://github.com/strukturag/libheif#software-using-libheif does not list ffmpeg as software which uses libheif.
> 
> On that basis, since MABS is a _build system_ for ffmpeg and dependencies, I guess that no change to MABS would achieve what you are seeking ?
> 
> Lack of heif support seems to be a long overdue shortcoming in ffmpeg, based on TRAC comments.

It's funny (or not) that ffmpeg supports so called "avif" but no "heic". It's just a container!
No big deal though, things are just starting to get complicated as some people already *removed* jpeg-xl.

**Comment by hydra3333** [comment]
Yes, things seem a tad peculiar. 
For better or worse, I'd always thought of ffmpeg as a mainstream video processing product and so was unsurprised to note some deficiencies around image processing.

**Comment by gitoss** [comment]
> hi, is it possible to add libheif to this script? https://github.com/strukturag/libheif

Using media-autobild's msys shell, I was able to compile libheif with all HEVC & AV1 encoders/decoders I want to use - there are no Windows binaries around the web for that. 

I didn't use add it to media-suite_compile.sh, but did a manual cmake - which was a LOT of hassle, I had to set the respective PKGCONF LIBRARY and INCLUDE_DIR params.

If anyone happens to have a patch / addition for media-suite_compile.sh which is 'fire and forget', please do share.

**Comment by LigH-de** [comment]
There is another request in the [German Gleitz video forum](https://gleitz.info/forum/index.php?thread/48600-heif-compilieren/&postID=467485#post467485)

**Comment by L4cache** [comment]
> I'm a bit late to the party. A new comment over on ffmpeg TRAC: https://trac.ffmpeg.org/ticket/6521#comment:23

That commit was not related to libheif. It's the native demuxer in FFmpeg. But still no muxer support for HEIC.
I'm thinking, how hard could it be?

**Comment by LigH-de** [comment]
And even as long as ffmpeg hesitates to implement new support for single image formats based on established video format I-frames, some people are interested in at least separate encoders for these formats.

**Comment by L4cache** [comment]
I'll see what I can do after it finished building everything with new version of clang ;)

**Comment by L4cache** [comment]
Test branch
https://github.com/L4cache/media-autobuild_suite/tree/test-libheif

I don't know what to do with the linking issues...
I'll open PR after at least the libde265 one is fixed...

**Comment by LigH-de** [comment]
Successfully built [libheif 1.19.3 8a139a6](https://www.mediafire.com/file/59ggbiqm56pyqoy/libheif_1.19.3_8a139a6.7z/file)

32 and 64 bit code contain a different selection of codecs; and heif-dec contains no HEIC decoder. (I wonder if libavcodec is able?)

**Comment by hydra3333** [comment]
Thank you.

Also, eek at 
> 32 and 64 bit code contain a different selection of codecs; and heif-dec contains no HEIC decoder. (I wonder if libavcodec is able?)

**Comment by LigH-de** [comment]
Well, of course, some codecs like VVC exist in 64 bit only. No surprise. The other issue sounds more important, if it is called "heif-dec", it should be able to fulfill its name at least. So hoping for libde265 or any alternative decoder to work,

**Comment by LigH-de** [comment]
I did not test daily. Now that I downloaded the latest release, VulkanLoader fails... makes testing hard until solved.

**Comment by zedd-0** [comment]
> 32 and 64 bit code contain a different selection of codecs; and heif-dec contains no HEIC decoder. (I wonder if libavcodec is able?)

There's something strange about that.
When I assemble the libheif decoder from the MSYS2 32-bit repository myself (precompiled binarys),
 it lists 'libde256' as the decoder, and it works too.

some links:
https://packages.msys2.org/packages/mingw-w64-i686-libheif
https://repo.msys2.org/mingw/i686/
https://repo.msys2.org/msys/i686/
(liblzma is unlisted and needs to be renamed.)

**Comment by LigH-de** [comment]
As I already mentioned: At that time when I tested it, this branch did not yet solve the linking issue.

Now it is fixed, libde265 links. But I could not yet test it again.

**Comment by LigH-de** [comment]
Disabling vulkan and libplacebo due to issues. And OpenH264 asks for a newer version because MSYS2 ships newer headers.

The following codecs are reported for my build:

### Win32
`heif-enc.exe --list-encoders`
```
AVC encoders:
AVIF encoders:
- aom = AOMedia Project AV1 Encoder 3.11.0-106-g80070c4b [default]
HEIC encoders:
- x265 = x265 HEVC encoder (4.0+61-7cc4030) [default]
- kvazaar = kvazaar HEVC encoder
JPEG encoders:
- jpeg = libjpeg-turbo 3.0.4 (libjpeg 8.0) [default]
JPEG 2000 encoders:
JPEG 2000 (HT) encoders:
Uncompressed encoders:
- uncompressed = builtin [default]
VVIC encoders:
```

`heif-dec.exe --list-decoders`
```
AVC decoders:
- openh264 = OpenH264 2.4.1
AVIF decoders:
- dav1d = dav1d v7.0.0
- aom = AOMedia Project AV1 Decoder 3.11.0-106-g80070c4b
HEIC decoders:
- libde265 = libde265 HEVC decoder, version 1.0.15
JPEG decoders:
- jpeg = libjpeg-turbo 3.0.4 (libjpeg 8.0)
JPEG 2000 decoders:
JPEG 2000 (HT) decoders:
uncompressed:
- uncompressed = builtin
VVIC decoders:
```

### Win64
`heif-enc.exe --list-encoders`
```
AVC encoders:
AVIF encoders:
- aom = AOMedia Project AV1 Encoder 3.11.0-106-g80070c4b [default]
- svt = SVT-AV1 encoder v2.3.0-56-g94a1a85
HEIC encoders:
- x265 = x265 HEVC encoder (4.0+61-7cc4030) [default]
- kvazaar = kvazaar HEVC encoder
JPEG encoders:
- jpeg = libjpeg-turbo 3.0.4 (libjpeg 8.0) [default]
JPEG 2000 encoders:
JPEG 2000 (HT) encoders:
Uncompressed encoders:
- uncompressed = builtin [default]
VVIC encoders:
- vvenc = vvenc VVC encoder [default]
- uvg266 = uvg266 VVC encoder
```

`heif-dec.exe --list-decoders`
```
AVC decoders:
- openh264 = OpenH264 2.4.1
AVIF decoders:
- dav1d = dav1d v7.0.0
- aom = AOMedia Project AV1 Decoder 3.11.0-106-g80070c4b
HEIC decoders:
- libde265 = libde265 HEVC decoder, version 1.0.15
JPEG decoders:
- jpeg = libjpeg-turbo 3.0.4 (libjpeg 8.0)
JPEG 2000 decoders:
JPEG 2000 (HT) decoders:
uncompressed:
- uncompressed = builtin
VVIC decoders:
- vvdec = VVCDEC decoder (3.0.0)
```

Could x264 or OpenH264 be used as AVC encoder?

**Comment by L4cache** [comment]
x264 support is not implemented in upstream. There's a commented line in CMakeLists refering OpenH264 encoding but it's not implemented as well. (it's in the TODO list, maybe?)
So... there's no any AVC encoding support in the moment.

**Comment by LigH-de** [comment]
You may have a little typo in the install script, a path is doubled:

`Error copying file "G:/MABS/local64/G:/MABS/local64/bin-video/heif-dec.exe" to "G:/MABS/local64/G:/MABS/local64/bin-video/heif-convert.exe".`

**Comment by L4cache** [comment]
Sometimes I wonder myself where did that heif-convert.exe go?
Thanks for pointing out, but that's something more weird than a typo:
I first noticed the heif-convert was gone after I set the install location to the /local64/bin-video folder instead of the /local64/bin folder.
After I realized it was just a link / copy (depending on OS) of heif-dec I just stopped bothering it.

https://github.com/strukturag/libheif/blob/2e02ee2b64941c3f693f6536d637926d450b97fb/examples/CMakeLists.txt#L31-L36

But this copy process is controled by the CMake, since the rest of the executables get installed in the right place... I don't think I did a typo. It must be CMake that did something wrong.

**Comment by LigH-de** [comment]
BTW: libde265 compiles to a separate dec265.exe too; but it does not print any console output, not even its own help text. Probably nothing to really worry about, as long as nobody seriously wants to use it...

**Comment by L4cache** [comment]
> BTW: libde265 compiles to a separate dec265.exe too; but it does not print any console output, not even its own help text. Probably nothing to really worry about, as long as nobody seriously wants to use it...

You ready? imma say it, here it goes: it's fine on my side.
You had similar issue with av1an as well, I guess it's something unusual with you system?

**Comment by LigH-de** [comment]
In case of Av1an, it was necessary to have Python 3.12+ and VapourSynth in it installed. 

Does dec265 have the same requirements? If so, is that documented?

**Comment by L4cache** [comment]
> In case of Av1an, it was necessary to have Python 3.12+ and VapourSynth in it installed.
> 
> Does dec265 have the same requirements? If so, is that documented?

I don't think so, the here built dec265's requirement should be just itself (and some standrad Windows libraries / runtimes).

**Comment by LigH-de** [comment]
My system is indeed a little special, it is Windows 10 upgraded from Windows 7. I noticed that the OS detection may behave slightly differently when running either as normal user or "as Administrator". In case of Stellarium, the Qt6 installer failed, only the Qt5 installer worked... But I see no reason why any console application would not want to print anything (possibly fail before any output and crash?). In a most recent build, Av1an [creates a log which looks very brief to me](https://github.com/master-of-zen/Av1an/issues/883#issuecomment-2524292470).

**Comment by lukenash707** [comment]
I'm a bit of a newbie when it comes to coding but I'm try to fumble my way through getting chat gpt to make a script to extract frames from a 10 bit video .mov using FFMPEG, I want to output 10 bit heic files but I can't seen to find a prebuilt binary for ffmpeg with heic/heif support is one out there or is this still in progress?

**Comment by lukenash707** [comment]
Yeah, unfortunately that would destroy the 10bit data which I’d like to keep, ive experimented with avif, jxl, png, and tiff but the decoder takes about 100x as long than working with jpeg! I think im just going to have to stick with 8 bit now until they add heif support

**Comment by L4cache** [comment]
Well, by design the HEVC, even with only one Intra frame, is much, much more complex than jpeg (1992).
Same thing can be said for other modern formats as well.

**Comment by LigH-de** [comment]
Please don't mind me reminding you that thisproject is not the perfect place for cozy discussions about the usage of tools, as mentioned in the readme of this project. Forums like doom9 or VideoHelp are more suitable.

**Comment by LigH-de** [comment]
@L4cache - trying to keep adding libheif to the current M-AB-S script fails with some strange errors. I guess their helper routines changed somewhat?

```
{...}
└ ffmpeg-git  .................................................. [Updated]
rm: unknown option -- g
Try 'rm --help' for more information.
rm: unknown option -- g
Try 'rm --help' for more information.
  Running git clone for  ...
 git seems to be down
Try again later or <Enter> to continue

if you're sure nothing depends on it.
rm: unknown option -- g
Try 'rm --help' for more information.
rm: unknown option -- g
Try 'rm --help' for more information.
  Running git clone for  ...
 git seems to be down
Try again later or <Enter> to continue

if you're sure nothing depends on it.
  Running svn update for mplayer...
{...}
```

**Comment by L4cache** [comment]
@LigH-de You can use the [PR branch](https://github.com/L4cache/media-autobuild_suite/tree/patch-9), replace [the link to the libheif patch](https://github.com/L4cache/media-autobuild_suite/blob/daeadef433488234c59893c2062fb8c11d98f79e/build/media-suite_compile.sh#L2479) to [its PR branch](https://github.com/L4cache/mabs-patches/tree/patch-1) as well

**Comment by LigH-de** [comment]
OK, I guess my main mistake was that I missed to maintain added deps in the ~deps.sh.

**Comment by LigH-de** [comment]
I guess libheif needs some similar [patch](https://github.com/m-ab-s/media-autobuild_suite/commit/3967ddcb559693b6891e7441c2c396d954e22200) as general ffmpeg to support SVT-AV1 v3.

```
┌ libheif git  ........................................ [Recently updated]
├ Running cmake...
├ Running build...
Likely error (tail of the failed operation logfile):
G:/MABS/build/libheif-git/libheif/plugins/encoder_svt.cc:716:16: error: 'EbSvtAv1EncConfiguration' {aka 'struct EbSvtAv1EncConfiguration'} has no member named 'high_dynamic_range_input'
  716 |     svt_config.high_dynamic_range_input = (bitdepth_y == 10 && // TODO: should this be >8 ?
      |                ^~~~~~~~~~~~~~~~~~~~~~~~
G:/MABS/build/libheif-git/libheif/plugins/encoder_svt.cc:722:16: error: 'EbSvtAv1EncConfiguration' {aka 'struct EbSvtAv1EncConfiguration'} has no member named 'color_description_present_flag'
  722 |     svt_config.color_description_present_flag = false;
      |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
G:/MABS/build/libheif-git/libheif/plugins/encoder_svt.cc:728:14: error: 'EbSvtAv1EncConfiguration' {aka 'struct EbSvtAv1EncConfiguration'} has no member named 'logical_processors'
  728 |   svt_config.logical_processors = encoder->threads;
      |              ^~~~~~~~~~~~~~~~~~
ninja: build stopped: subcommand failed.
build failed. Check G:/MABS/build/libheif-git/build-64bit/ab-suite.build.log
This is required for other packages, so this script will exit.
```
